### PR TITLE
Move etcd upgrade code into role

### PR DESCRIPTION
**Pull request by ingvagabund**
The is a preparation for merging etcd upgrade related code with etcd role.
This just basically moves code from etcd upgrade playbooks into a role wherever it makes sense + minor refactoring.

### CODE REVIEW

**Comment by ingvagabund** [comment]
aos-ci-test

File: playbooks/common/openshift-cluster/redeploy-certificates/ca.yml [hunk 1]
```diff
@@ -9,7 +9,8 @@
 - name: Backup existing etcd CA certificate directories
   hosts: oo_etcd_to_config
   roles:
-  - etcd_common
+  - role: etcd_common
+    r_etcd_common_etcd_runtime: "{{ 'runc' if openshift.common.is_etcd_system_container | bool else 'docker' if openshift.common.is_containerized | bool else 'host' }}"
```
Comment (ashcrow) [inline playbooks/common/openshift-cluster/redeploy-certificates/ca.yml]: This line is used a lot. Is there a way to set it once and use it? Or does it need to be created each time?

Comment (ingvagabund) [inline playbooks/common/openshift-cluster/redeploy-certificates/ca.yml]: Yeah, it bothers me as well. Maybe move it under openshift_etcd_facts as:
``etcd_runtime: "{{ 'runc' if openshift.common.is_etcd_system_container | bool else 'docker' if openshift.common.is_containerized | bool else 'host' }}"`` and then used it here.

Comment (ashcrow) [inline playbooks/common/openshift-cluster/redeploy-certificates/ca.yml]: That sound good to me!

Comment (ingvagabund) [inline playbooks/common/openshift-cluster/redeploy-certificates/ca.yml]: Found a spot in `openshift_facts` role to set `etcd_runtime`.

File: playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml [hunk 2]
```diff
@@ -2,118 +2,109 @@
...
-    # AUDIT:changed_when: `false` because we are only inspecting
-    # state, not manipulating anything
-    changed_when: false
-
-# I really dislike this copy/pasta but I wasn't able to find a way to get it to loop
-# through hosts, then loop through tasks only when appropriate
-- name: Upgrade to 2.1
-  hosts: oo_etcd_hosts_to_upgrade
-  serial: 1
+  - block:
```
Comment (ewolinetz) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: Would this eventually be able to be within the `etcd` role once it is all integrated? Or can we move this into the `etcd_upgrade` role too?

Comment (ingvagabund) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: I would rather see this as part of a different role. E.g. 'version_detection` or similar one. Given each task is a version detection, calling an `etcd/etcd_upgrade` role to get it is too expensive.

**Comment by ingvagabund** [comment]
Given the etcd members need to be all updated to version+1 until they are updated to version+2, the `etcd_upgrade` gets called (CLUSTER_SIZE * NUMBER_OF_UPDATES)-times. So it is important to decrease the number of dependencies of etcd roles (for both rpm and image based) so e.g. the docker role does not get called every time a member is updated to a newer image (since once we do the first update, it is likely the docker is already there and properly configured when updating for the second time)

**Comment by ingvagabund** [comment]
aos-ci-test

**Comment by ingvagabund** [comment]
@sdodson I would like to have this merged before the 1.6 branch splits. It upgrades etcd to 3.1 as well.

File: playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml [hunk 3]
```diff
@@ -54,84 +54,57 @@
...
-    when: etcd_container_version | default('99') | version_compare('3.0','<') and openshift.common.is_containerized | bool
+    etcd_upgrade_version: '3.0.15'
+
+- include: upgrade_rpm_members.yml
+  vars:
+    etcd_upgrade_version: '3.1'
+
+- include: upgrade_image_members.yml
+  vars:
+    etcd_upgrade_version: '3.1.3'
```
Comment (ingvagabund) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: @sdodson Given the latest rpm etcd is 3.1.* and the latest 3.1 image is 3.1.3, it used those two values.

Comment (sdodson) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: sounds good

**Comment by sdodson** [comment]
This looks really good, going to run through it locally since CI tests don't perform an upgrade.

File: playbooks/common/openshift-cluster/redeploy-certificates/etcd.yml [hunk 4]
```diff
@@ -50,6 +52,7 @@
       etcd_peers: "{{ groups.oo_etcd_to_config | default([], true) }}"
       etcd_certificates_etcd_hosts: "{{ groups.oo_etcd_to_config | default([], true) }}"
       openshift_ca_host: "{{ groups.oo_first_master.0 }}"
+      r_etcd_common_etcd_runtime: "{{ openshift.common.etcd_runtime }}"
```
Comment (mtnbikenc) [inline playbooks/common/openshift-cluster/redeploy-certificates/etcd.yml]: I see where this is headed.  I don't think we should be passing in role variables which will be used by other (dependency) roles.  This is where we need to flatten out our role dependencies first it seems.  The only vars we should be passing in are ones for the role we are calling.

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml [hunk 5]
```diff
@@ -1,84 +1,14 @@
...
-  - name: Display location of etcd backup
-    debug:
-      msg: "Etcd backup created in {{ etcd_backup_dir }}"
+  - role: openshift_facts
+  - role: etcd_upgrade
+    r_etcd_upgrade_action: backup
+    r_etcd_backup_tag: etcd_backup_tag
+    r_etcd_common_etcd_runtime: "{{ openshift.common.etcd_runtime }}"
+    r_etcd_upgrade_embedded_etcd: "{{ groups.oo_etcd_to_config | default([]) | length == 0 }}"
+    r_etcd_backup_sufix_name: "{{ lookup('pipe', 'date +%Y%m%d%H%M%S') }}"
```
Comment (mtnbikenc) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: Again here vars are being passed in which are not for the role being called.

File: playbooks/common/openshift-cluster/redeploy-certificates/ca.yml [hunk 6]
```diff
@@ -52,7 +53,8 @@
   vars:
     etcd_ca_host: "{{ groups.oo_etcd_to_config.0 }}"
   roles:
-  - etcd_common
+  - role: etcd_common
+    r_etcd_common_etcd_runtime: "{{ openshift.common.etcd_runtime }}"
```
Comment (ashcrow) [inline playbooks/common/openshift-cluster/redeploy-certificates/ca.yml]: 👍 on ``r_``

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml [hunk 7]
```diff
@@ -1,84 +1,14 @@
...
-  - set_fact:
-      etcd_backup_complete: True
-
-  - name: Display location of etcd backup
-    debug:
-      msg: "Etcd backup created in {{ etcd_backup_dir }}"
+  - role: openshift_facts
+  - role: etcd_upgrade
+    r_etcd_upgrade_action: backup
+    r_etcd_backup_tag: etcd_backup_tag
```
Comment (mtnbikenc) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: What is `etcd_backup_tag`? global? inventory?

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml [hunk 8]
```diff
@@ -88,7 +18,7 @@
   - set_fact:
       etcd_backup_completed: "{{ hostvars
                                  | oo_select_keys(groups.oo_etcd_hosts_to_backup)
-                                 | oo_collect('inventory_hostname', {'etcd_backup_complete': true}) }}"
+                                 | oo_collect('inventory_hostname', {'r_etcd_upgrade_backup_complete': true}) }}"
```
Comment (mtnbikenc) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: I don't think using a role var here is the right thing to do.

File: playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml [hunk 9]
```diff
@@ -54,84 +54,57 @@
...
 
 - name: Upgrade fedora to latest
   hosts: oo_etcd_hosts_to_upgrade
   serial: 1
   tasks:
-  - include: fedora_tasks.yml
-    when: ansible_distribution == 'Fedora' and not openshift.common.is_containerized | bool
+  - include_role:
```
Comment (mtnbikenc) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: Why does this have to be `include_role` and not just moved into a `roles:` stage?

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml (hunk 5 above)
Comment (ashcrow) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: nit: Not a required change, but you can use ``ansible_date_time.iso8601`` to get "2017-05-19T18:45:08Z". If the ``-``'s are a problem then ``ansible_date_time.iso8601|replace('-', '')``.

File: playbooks/common/openshift-cluster/upgrades/etcd/main.yml [hunk 10]
```diff
@@ -8,7 +8,7 @@
 - name: Backup etcd before upgrading anything
   include: backup.yml
   vars:
-    backup_tag: "pre-upgrade-"
+    etcd_backup_tag: "pre-upgrade-"
```
Comment (ashcrow) [inline playbooks/common/openshift-cluster/upgrades/etcd/main.yml]: 👍 on naming.

File: playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml [hunk 11]
```diff
@@ -12,10 +12,10 @@
       # AUDIT:changed_when: `false` because we are only inspecting
       # state, not manipulating anything
       changed_when: false
-
     - debug:
         msg: "Etcd rpm version {{ etcd_rpm_version.stdout }} detected"
-    when: not openshift.common.is_containerized | bool
+    when:
+    - not openshift.common.is_containerized | bool
```
Comment (ashcrow) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: 👍 on multiline when.

File: roles/etcd_upgrade/meta/main.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+---
+galaxy_info:
+  author: Jan Chaloupka
+  description:
+  company: Red Hat, Inc.
+  license: Apache License, Version 2.0
+  min_ansible_version: 1.9
```
Comment (ashcrow) [inline roles/etcd_upgrade/meta/main.yml]: Are you sure ``1.9`` can cut it?

File: roles/etcd_upgrade/tasks/backup.yml [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  # AUDIT:changed_when: `false` because we are only inspecting
+  # state, not manipulating anything
+  changed_when: false
+
+- name: Abort if insufficient disk space for etcd backup
+  fail:
+    msg: >
+      {{ etcd_disk_usage.stdout }} Kb disk space required for etcd backup,
+      {{ avail_disk.stdout }} Kb available.
+  when: (r_etcd_upgrade_embedded_etcd | bool) and (etcd_disk_usage.stdout|int > avail_disk.stdout|int)
```
Comment (ashcrow) [inline roles/etcd_upgrade/tasks/backup.yml]: nit: Not required but multiline would be perfect here 😄

**Comment by ashcrow** [comment]
@mtnbikenc are these things you think need to be done before this feature can be merged, or would it be acceptable to merge and have @ingvagabund follow on cleaning up items to meet our new specs?

**Comment by mtnbikenc** [comment]
@ashcrow I'm ok with moving forward with these changes, however, I don't want to establish a pattern that is against what we are attempting to do with role variables.  I think there are a few architectural things we need to discuss/decide/document so we can make sure we are all going in the same direction.

**Comment by ashcrow** [comment]
@mtnbikenc that makes sense to me.

@sdodson deferring to you for merge

File: roles/etcd_upgrade/meta/main.yml (hunk 12 above)
Comment (ingvagabund) [inline roles/etcd_upgrade/meta/main.yml]: We need to update all meta/main.yml and bump all `min_ansible_version` to the same or run ansible's check.

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml (hunk 7 above)
Comment (ingvagabund) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: playbook var set under upgrades/etcd/* playbooks

File: playbooks/common/openshift-cluster/redeploy-certificates/etcd.yml (hunk 4 above)
Comment (ingvagabund) [inline playbooks/common/openshift-cluster/redeploy-certificates/etcd.yml]: The more we talk about the more it should be `r_etcd_upgrade_etcd_runtime` which sets its `r_etcd_common_etcd_runtime` equivalent in `etcd_upgrade` role. This is a first step before I finish refactoring all etcd like roles. I will take care of it.

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml (hunk 5 above)
Comment (ingvagabund) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: I did not renamed all variables, will cover it later.

File: playbooks/common/openshift-cluster/upgrades/etcd/backup.yml (hunk 8 above)
Comment (ingvagabund) [inline playbooks/common/openshift-cluster/upgrades/etcd/backup.yml]: Maybe `o_etcd_upgrade_back_complete` to state the variable is outputed by `etcd_upgrade` role?

File: playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml (hunk 9 above)
Comment (ingvagabund) [inline playbooks/common/openshift-cluster/upgrades/etcd/upgrade.yml]: Since I forget to replace it with the `role`.

**Comment by ingvagabund** [comment]
@mtnbikenc I agree with what you say. This is a middle step before the role is fully conforming to what we agreed and will agree on arch meetings. I have been basically cleaning up etcd related tasks and this is the result. I am continue with additional refactoring in other PR in which I will merge both `etcd` and `etcd_upgrade` roles into one.

**Comment by ingvagabund** [comment]
@mtnbikenc moving forward and merging

**Comment by ingvagabund** [comment]
[merge]

**Comment by ingvagabund** [comment]
```sh
Ran 26 of 849 Specs in 852.087 seconds
SUCCESS! -- 26 Passed | 0 Failed | 2 Pending | 821 Skipped May 20 18:34:42.469: INFO: Error running cluster/log-dump.sh: fork/exec /data/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/cluster/log-dump.sh: no such file or directory
PASS

Ginkgo ran 1 suite in 14m12.510826685s
Test Suite Passed
[INFO] [CLEANUP] Beginning cleanup routines...
[INFO] [CLEANUP] Dumping cluster events to _output/scripts/conformance/artifacts/events.txt
[INFO] [CLEANUP] Dumping container logs to _output/scripts/conformance/logs/containers
[INFO] [CLEANUP] Truncating log files over 100M
[INFO] [CLEANUP] Stopping docker containers
[INFO] [CLEANUP] Removing docker containers
[INFO] [CLEANUP] Killing child processes
[INFO] [CLEANUP] Pruning etcd data directory
rm: cannot remove ‘/tmp/etcd/member’: Permission denied
[ERROR] test/extended/conformance.sh exited with code 1 after 00h 43m 37s
make: *** [test-extended] Error 1
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
########## FINISHED STAGE: FAILURE: RUN EXTENDED TESTS ##########
```

Two think
- github.com/openshift/origin/vendor/k8s.io/kubernetes/cluster/log-dump.sh: based on commit history the file never existed, strange
- `rm: cannot remove ‘/tmp/etcd/member’: Permission denied`: this should not cause a failure even if the directory can not be removed. Assuming the VM is destroyed, this is no issue.

**Comment by ingvagabund** [comment]
@jhadvig ^^

**Pull request merged**